include/bucket.h

```diff
@@ -2,76 +2,93 @@
    */
-  inline static ubucket_t gen_r(const XXH64_hash_t& bucket_seed, const ubucket_t& large_prime) {
-    return 2 + bucket_seed % (large_prime - 3);
-  }
+  inline static vec_t index_hash(const vec_t& index, long seed);
 
   /**
    * Checks whether the hash associated with the Bucket hashes the index to 0.
-   * @param index
-   * @param bucket_seed
-   * @param guess_nonzero
-   * @return true if the index is NOT hashed to zero.
+   * @param col_index_hash The return value to Bucket::col_index_hash
+   * @param guess_nonzero A power of 2, used as a modulus
+   * @return true if the index is NOT hashed to zero mod guess_nonzero.
    */
-  bool contains(const vec_t& index, const XXH64_hash_t& bucket_seed, const vec_t& guess_nonzero) const;
+  inline bool contains(const vec_t& col_index_hash, const vec_t& guess_nonzero) const;
 
   /**
    * Checks whether this Bucket is good.
    * @param n Size of the vector being sketched.
-   * @param large_prime Modulus to use in c caluclation.
-   * @param bucket_seed This Bucket's seed, generated with gen_bucket_seed.
-   * @param r This Bucket's r, generated with gen_r.
+   * @param bucket_col This Bucket's column index.
    * @param guess_nonzero The guess of nonzero elements in the vector being sketched.
-   * @return true if this bucket is good, else false.
-   */
-  bool is_good(const vec_t& n, const ubucket_t& large_prime, const XXH64_hash_t& bucket_seed, const ubucket_t& r, const vec_t& guess_nonzero) const;
-
-  /**
-   * Updates this Bucket with the given Update
-   * @param update
-   * @param large_prime Modulus to use in c caluclation.
-   * @param r This Bucket's r, generated with gen_r.
+   * @param sketch_seed The seed of the Sketch this Bucket belongs to.
+   * @return true if this Bucket is good, else false.
    */
-  void update(const Update& update, const ubucket_t& large_prime, const ubucket_t& r);
+  inline bool is_good(const vec_t& n, const unsigned bucket_col, const vec_t& guess_nonzero, const long& sketch_seed) const;
 
   /**
-   * Updates this Bucket with the given Update, using cached powers of r
-   * @param update
-   * @param large_prime Modulus to use in c caluclation.
-   * @param r_sq_cache A vector, where r_sq_cache[i] = r^2^i. Generated with
-   * PrimeGenerator::gen_sq_cache
+   * Updates this Bucket with the given update index
+   * @param update_idx The update index
+   * @param update_hash The hash of the update index, generated with Bucket::index_hash.
    */
-  void cached_update(const Update& update, const ubucket_t& large_prime, const std::vector<ubucket_t>& r_sq_cache);
+  inline void update(const vec_t& update_idx, const vec_t& update_hash);
 };
 
+inline vec_t Bucket_Boruvka::col_index_hash(const unsigned bucket_col, const vec_t& update_idx, const long sketch_seed) {
+  struct {
+    unsigned bucket_col;
+    vec_t update_idx;
+  } __attribute__((packed)) buf = {bucket_col, update_idx};
+  return
+#ifdef USE_NATIVE_F
+  XXH32
+#else
+  XXH3_64bits_withSeed
```

Comment:
what's the difference between XXH32 and XXH3?

Reply:
Here's the benchmark table from XXH's page:
| Hash Name     | Width | Bandwidth (GB/s) | Small Data Velocity | Quality | Comment |
| ---------     | ----- | ---------------- | ----- | --- | --- |
| __XXH3__ (SSE2) |  64 | 31.5 GB/s        | 133.1 | 10
| __XXH128__ (SSE2) | 128 | 29.6 GB/s      | 118.1 | 10
| _RAM sequential read_ | N/A | 28.0 GB/s  |   N/A | N/A | _for reference_
| City64        |    64 | 22.0 GB/s        |  76.6 | 10
| T1ha2         |    64 | 22.0 GB/s        |  99.0 |  9 | Slightly worse [collisions]
| City128       |   128 | 21.7 GB/s        |  57.7 | 10
| __XXH64__     |    64 | 19.4 GB/s        |  71.0 | 10
| SpookyHash    |    64 | 19.3 GB/s        |  53.2 | 10
| Mum           |    64 | 18.0 GB/s        |  67.0 |  9 | Slightly worse [collisions]
| __XXH32__     |    32 |  9.7 GB/s        |  71.9 | 10
| City32        |    32 |  9.1 GB/s        |  66.0 | 10
| Murmur3       |    32 |  3.9 GB/s        |  56.1 | 10
| SipHash       |    64 |  3.0 GB/s        |  43.2 | 10
| FNV64         |    64 |  1.2 GB/s        |  62.7 |  5 | Poor avalanche properties
| Blake2        |   256 |  1.1 GB/s        |   5.1 | 10 | Cryptographic
| SHA1          |   160 |  0.8 GB/s        |   5.6 | 10 | Cryptographic but broken
| MD5           |   128 |  0.6 GB/s        |   7.8 | 10 | Cryptographic but broken

So for "native" we're using 32 bit integers, so may as well use the 32 bit hash XXH32.
For 64 bit integers, we were using XXH64, but it turns out XXH3 is just better in every way.